### ISSUE TITLE
Wrong environment variables (QT_QPA_PLATFORMTHEME=qt5ct not set)

### ISSUE THREAD

**Issue by matheusfillipe** [Ulauncher/Ulauncher#780]
## Problem 

When I launch an application from ulauncher seems that some environment variables are lost.

I have `QT_QPA_PLATFORMTHEME=qt5ct` on my /etc/environment, ~/.profile, ~/.zshrc and tried launching ulauncher with ``QT_QPA_PLATFORMTHEME=qt5ct ulauncher``

Still when I launch ulauncher it is different from launching the application from command line (where the variable is set).

### From ulauncher
![image](https://user-images.githubusercontent.com/24435787/129811929-af71df33-314f-45a8-a6b7-7b70023cfa14.png)

### From Terminal
![image](https://user-images.githubusercontent.com/24435787/129811987-a0bb6bd6-c688-434f-a56d-b1646c6d9855.png)
(The theme and font are very different)

## Environment

I am on qtile window manager, a very minimal and simple window manager. This problem doesn't happen for me (same OS) on KDE plasma.  I am building ulauncher with this little script based on the PKGBUILD:

```sh
build() {
  python setup.py build
  ./ul build-preferences
}

prepare() {
  pkgver=$(git describe --long --tags | sed -r 's/([^-]*-g)/r\1/;s/-/./g')
  sed -i "s/%VERSION%/${pkgver%.*.*}/g" setup.py
  find -iname "*.py" | xargs sed -i 's=\(^#! */usr/bin.*\)python3 *$=\1python='
}

package() {
  export PYTHONHASHSEED=0
  sudo mv build/share/applications/ulauncher.desktop /usr/share/applications
  sudo python setup.py install --root="/" --optimize=1 --skip-build
  sudo rm -rf /usr/share/ulauncher/preferences/{no*,src,bow*,gul*,pack*}
  sudo mv ulauncher.service /usr/lib/systemd/user/
}

prepare
build
package

```

This problem also happened with qtile from aur. `qtile-git`

OS: Arch Linux x86_64, CPU: Intel i7-7700HQ (8) @ 3.800GHz, GPU: NVIDIA GeForce GTX 1050 Ti Mobile, GPU: Intel HD Graphics 630, Memory: 15.5 GiB

**Comment by friday** [comment]
Which Ulauncher version? You're building from the dev branch? We don't have a stable release yet, but in the process of releasing one including #778, which could be causing this?

**Comment by matheusfillipe** [comment]
Yes I am building from dev. 
I also just went all the way back to 5.10 to test it and I still have the issue. If I launch ulauncher from a terminal with `QT_QPA_PLATFORMTHEME` set the qt programs don't see this variable and have the wrong theme. If I stop ulauncher and open them from the same terminal they look ok.

I am not sure if it could be something on my system but doesn't make sense. Is ulauncher doing something with those environmental variables somewhere? I couldn't find nothing that could be the cause it so far.

**Comment by friday** [comment]
Systemd doesn't source your home environment. If you build from the new issue-780 branch I just pushed, and then run ulauncher directly from the terminal (`systemctl --user stop ulauncher` then `ulauncher -v`), does it work?

**Comment by matheusfillipe** [comment]
I am launching from the terminal as I said, not systemd. But i do have the systemd daemon.... let me see.

Hah! After I stopped the systemd ulauncher it works! Sorry my mistake, I didn't realize I had it enabled and I didn't think that with it enabled the `ulauncher` command was actually starting it.

Sorry for the dumb issue :P!

**Comment by friday** [comment]
Not a dumb issue at all :) I am going to include the fix from that branch in the next release because of this issue, which should fix cases like this when not running ulauncher from systemd. And maybe we should document this caveat when running from systemd.

**Comment by friday** [comment]
The commit is in the main branch now and I deleted the the issue-780 branch.

**Comment by matheusfillipe** [comment]
Nice! I just pulled and tested it! It works with the systemd daemon now too!

**Comment by friday** [comment]
It shouldn't actually.

This creates a desktop entry (app) which shows your environment variables:

```sh
printf '[Desktop Entry]                
Name=Printenv
Comment=Show environment variables
Exec=sh -c "printenv | less -"
Icon=text-x-script
Terminal=true
Type=Application' > ~/.local/share/applications/printenv.desktop
```

When I run that from the gnome shell launcher (Super+A) or Ulauncher started via systemd it says my editor variable is is `EDITOR=vim`, but it's actually `EDITOR=nvim`. Running ulauncher directly from the active shell prints the right variables. I think this is all we can do though, other than document a workaround.

**Comment by matheusfillipe** [comment]
Yeah I think there isn't a better solution for this.

Maybe when ulauncher is first run from terminal, or with some option, at build, i don't know. It could generate an environment file and you can have the systemd service to load it with [https://kinvolk.io/docs/flatcar-container-linux/latest/setup/systemd/environment-variables/#environmentfile-directive](EnvironmentFile)

**Comment by troycurtisjr** [comment]
In this case ulauncher is simply passing along the environment that it is given. Unfortunately, how you define that environment depends. There is the well known shell environment places, like `~/.bash_profile` and `~/.profile`. X11 seems it would look at `~/.profile` as well. With systemd services they are defined in the service file itself using `Environment` or `EnvironmentFile`. Login sessions using pam, presuming `pam_env` is configured, look to `/etc/environment` and `~/.pam_environment`. systemd uses `/etc/environment.d/*.conf` and maybe also `/etc/environment` if the standard (but not required) symlink is in place. 

So all this to say that setting environment variables for sessions are really dependent on the environment, and kind of a mess tbh.  I suspect that now that you have the value in `/etc/environment`, and you've restarted the ulauncher systemd service, systemd correctly parsed the environment out of your file, and thus it is present in the ulauncher's environment now. Even if ulauncher wanted to try to setup the environment independent of the session it was started in, it would have to arbitrarily pick one of the methods listed above, likely breaking the other methods. I suspect the best thing for ulauncher is to just faithfully pass through the environment it is given, and the surrounding session needs to be responsible for setting the values, which it would need to be for any other application anyway.

**Comment by friday** [comment]
You can actually do `systemctl --user edit ulauncher` and manually make user-level changes/additions to the unit. I tried to make it set the `EDITOR` to nano this way, but it still wouldn't work (and yes I restarted afterwards).

I would not add functionality to manage environment variables to Ulauncher. This is a niche case. If you need them, then you can hopefully override the ulauncher systemd unit like above to make it work (even if I couldn't), or override the desktop entries themselves. The latter definitely works.

**Comment by friday** [comment]
If you want Ulauncher to have specific environment variables set now you can run

`systemctl --user edit ulauncher`, then add the variable in the designated area:

![2021-09-04_05-58](https://user-images.githubusercontent.com/515120/132081694-c9e57659-1e7b-4bbd-b17c-763f3e03b236.png)

Then save and run `systemctl --user restart ulauncher`. Then it should work :)

(although I think it's still better to add the environment variables to the desktop entries.

**Comment by darkfishy** [comment]
This issue (as reported originally by Matheus) seems to persist with version 5.14.1-1 on Arch/Sway. 

### Setup 1
1. Not using `systemd`. 
2. `ulauncher` is started from either a terminal (`Alacritty`) or via `sway` config (`exec ulauncher --hide-window`) at startup.

Irrespective of how `ulauncher` is started (point 2 above), certain environment variables are not available to apps started from ulauncher. However, when those apps are started directly from terminal or via `sway` config, they have those environment variables available.

If I understand correctly, `ulauncher` should pass on variables that were available to it at launch, which isn't the case. 

Logs attached: https://gist.github.com/darkfishy/7b9971e845b9fc0377138f19d89dcf47

### Setup 2
Start `ulauncher` via systemd with Environment specified using overrides. Even then the environment variables are not passed onto the applications started via `ulauncher`.


The reason I bring this to attention is because prior to v 5.14.1, it would work as expected.

**Comment by darkfishy** [comment]
The regression seems to be due to changes brought in commit 93c68de2b4c994840e577bb5a3aa5a5659da287b which has to do with the following lines:
https://github.com/Ulauncher/Ulauncher/blob/c1f570793401b57b412cbec7d1fd035c2b21f0e7/ulauncher/api/shared/action/LaunchAppAction.py#L53-L58

When an app is executed via GLib in https://github.com/Ulauncher/Ulauncher/blob/c1f570793401b57b412cbec7d1fd035c2b21f0e7/ulauncher/api/shared/action/LaunchAppAction.py#L68-L74

the current environment is not being passed onto the application being executed. This is due to the nature of `systemd-run` and the `envp` parameter to `GLib.spawn_async` has no effect on that.

I have a solution for this, though it would require further testing (more than my limited run).
Would a PR be welcome and accepted?

**Comment by friday** [comment]
I can reproduce the issue. We would be happy for a PR, but can't promise it would be merged. That depends on the quality of the PR. You seem like you know what you're doing, so I think we would.

**Comment by friday** [comment]
Seems like you can pass multiple `--setenv=X=Y` to `systemd-run`, so we can probably make it work that way. Was that what you had in mind? :)

**Comment by friday** [comment]
I made a PR @darkfishy ^. If you have a better solution in mind please let me know. I think this is kind of ugly, but whatever solves the issue.

**Comment by darkfishy** [comment]
> Seems like you can pass multiple `--setenv=X=Y` to `systemd-run`, so we can probably make it work that way. Was that what you had in mind? :)

Hey @friday That's exactly what I am doing, functionally. 😄 Though in my case, I had also moved the `envp` listcomp outside and before `if hasSystemdRun` and then reuse `envp` in another `setenvs` listcomp like you. 

There is, however, a caveat with this approach (as I found out while testing). It has to do with [quoting policy](https://www.man7.org/linux/man-pages/man7/systemd.syntax.7.html#QUOTING) of `systemd` and likely `GLib.spawn_async`. To illustrate, consider following scenarios.

1. User has following in their environment (somewhere, via `/etc/environment`, `~/.profile`, etc.)
```
export QT_QPA_PLATFORM='wayland;xcb'
export QT_QPA_PLATFORMTHEME=qt5ct
```

2. User has following lines in their `override.conf` for `ulauncher.service`
```
[Service]
Environment="QT_QPA_PLATFORM='wayland;xcb'"
Environment="QT_QPA_PLATFORMTHEME=qt5ct"
```
Note the nested single quotes `'`. This becomes important.

If `ulauncher` is started from a terminal or via DE/WM, say `exec` statements in `sway` config, this isn't a problem because shell has removed those single quotes already. So what `ulauncher` and applications started by `ulauncher` see is just `QT_QPA_PLATFORM=wayland;xcb`. 

However, if `systemd` is used to start `ulauncher`, then both it and other applications thereon see it as `QT_QPA_PLATFORM='wayland;xcb'`. 

This can be tested easily using the `printenv` desktop entry you have created above.

It may not be obvious to users, like me, but when starting an application, say `qt5ct`, Qt tries to look for `'wayland` and `xcb'` as platform plugins (which are invalid of course) and then the application fails to start.

This is by no means an issue with `ulauncher`, however, it may be prudent to warn users of this behaviour in the README or wiki somewhere.

What do you think?

**Comment by darkfishy** [comment]
@friday An alternative could be what I suggested in the PR https://github.com/Ulauncher/Ulauncher/pull/852#discussion_r747971443

**Comment by friday** [comment]
I forgot to close this issue again (it was fixed in 5.14.2)
